docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+
+- Maintains an in-memory collection of upkeepResults. It should maintain a single result per `upkeepTriggerID`. 
+    - Overwrites results for duplicated `upkeepTriggerID`.
+- Each result has a TTL. When the TTL expires
+    - conditional upkeeps will be sampled and checked again automatically
+    - log upkeeps that were missed are expected to be picked up by the log recoverer **TBD**
+- Provides a read API to view results, which takes as input (`pseudoRandomSeed`, `limit`). Sorts all keys (trigger, upkeepID) with the `seed` and provides first `limit` results. We do not do FIFO here as potentially out of sync old results will block new results sent by the node for consensus.
+- Provides an API to remove (trigger, upkeepID) from the store
+
+### Metadata Store
+
+This component stores the metadata to be sent within observations to the network. There are three categories of metadata
+
+- Conditional Upkeep Instructions
+- Log recovery Instructions
+- Latest block history
+
+Every item has a TTL, and upon expiry items are just removed without any action.
+
+The store provides an add / view / remove API for other components.
+
+### Samples Observer
+
+Processes samples of upkeeps to be checked, using the latest block number as a trigger. It does the following procedures:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **metadata store**
+with conditional Upkeep instructions
+- Errors and ineligible results are ignored
+
+### Conditional Observer
+
+Processes coordinated block + upkeep payload coming from plugin and does the following:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **result store**
+- Errors and ineligible results are ignored
+
+### Log Observer
+
+Processes latest logs for active upkeeps. It does the following procedures:
+
+- Called every second by Log Trigger Ticker with latest logs as full payloads.
+- Filter out logs with block number that is older than threshold
+    - These are expected to be picked up by the log recoverer 
+- Pre-processes to filter the logs already present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into result store
+- If runner gave an error, put back into log retry ticker, to be performed after a fixed timeout (can extend to exponential backoff)
+
+<aside>
+💡 Note: Duplicate logs can enter this flow, this component should be able to handle that. This component can maintain in-memory state for processed logs and can also use that state for exponential backoff
+</aside>
+
+### Retry Observer
+
+**TODO**
+
+### Recovery Observer
+
+This flow gets as input an upkeepID and a logIdentifier. It is a backup flow which verifies that a log was fully processed. Logs can be put here from the node itself (through Log Recoverer) or from other nodes (through Plugin)
```

Comment:
There are two parts to recovery:
- recovery proposal
- recovery with coordinated block

The first, `recovery proposal`, collects recoverable items from the node itself and runs a retry attempt cadence similar to the retry observer. The sources of recoverables in this component are
- recoverable provider
- result store TTL
- retry observer failures
- main log flow if retry observer rejects a result
- final recovery with coordinated block

No recovery proposals run through the pipeline during their retry cadence. Instead they are surfaced to the network and achieve quorum per recoverable. For each item added to this flow, a wait time is applied before attempting to put it into the metadata store. This wait time is once per minute with a maximum of 24 hours. Items can be added multiple times to the recovery proposal flow from multiple sources without altering wait times. This allows the recovery proposal to accept repeated proposals and provide a unified output.

The second, `recovery with coordinated block`, has an input of recoverable proposals that have achieved quorum, applies the latest coordinated block upon entry to the flow and are passed through the runner. If the results are eligible, they are forwarded to the result store. Otherwise, the results are sent back to the recovery proposal flow.

Reply:
Not fully understand why we need wait time before putting into metadata store

Reply:
The sources of recoverables can also be simplified by only having the log recoverer responsible for all recovery by keeping a strict range. Will discuss with @amirylm